src/main/scala/firrtl/options/StageAnnotations.scala

```diff
@@ -14,6 +17,63 @@ sealed trait StageOption { this: Annotation => }
   */
 trait Unserializable { this: Annotation => }
 
+/** Mix-in that lets an [[firrtl.annotations.Annotation Annotation]] serialize itself to a file separate from the output
+  * annotation file.
+  *
+  * This can be used as a mechanism to serialize an [[firrtl.options.Unserializable Unserializable]] annotation or to
+  * write ancillary collateral used by downstream tooling, e.g., a TCL script or an FPGA constraints file. Any
+  * annotations using this mix-in will be serialized by the [[firrtl.options.phases.WriteOutputAnnotations
+  * WriteOutputAnnotations]] phase. This is one of the last phases common to all [[firrtl.options.Stage Stages]] and
+  * should not have to be called/included manually.
+  *
+  * Note: from the perspective of transforms generating annotations that mix-in this trait, the serialized files are not
+  * expected to be available to downstream transforms. Communication of information between transforms must occur
+  * through the annotations that will eventually be serialized to files.
+  */
+trait CustomFileEmission { this: Annotation =>
+
+  /** Output filename where serialized content will be written
+    *
+    * The full annotation sequence is a parameter to allow for the location where this annotation will be serialized to
+    * be a function of other annotations, e.g., if the location where information is written is controlled by a separate
+    * file location annotation.
+    *
+    * @param annotations the annotation sequence at the time of emission
+    */
+  protected def baseFileName(annotations: AnnotationSeq): String
+
+  /** Optional suffix of the output file */
+  protected def suffix: Option[String]
+
+  /** A method that can convert this annotation to bytes that will be written to a file.
+    *
+    * If you only need to serialize a string, you can use the `getBytes` method:
+    * {{{
+    *  def toBytes: Option[Iterable[Byte]] = Some(myString.getBytes)
+    * }}}
+    */
+  def toBytes: Option[Iterable[Byte]]
+
+  /** Optionally, a sequence of annotations that will replace this annotation in the output annotation file.
```

Comment:
"that will replace this annotation **after serialization**" ?

Reply:
I actually think this is correct as written. This only replaces the annotation in the output annotation file. It's not replaced in the actual annotation sequence.

Put differently, the `WriteOutputAnnotations` phase is an identity function.

Maybe I'm misunderstanding?

Reply:
I was misunderstanding what WriteOutputAnnotations was doing. Yeah, it seems like it it correct as is.